docker/main/ngen/Dockerfile

```
@@ -290,14 +292,15 @@ RUN if [[ "${NETCDF_CXX_VERSION}" == "latest" ]]; then \
 FROM rocky-base as rocky-ngen-packaged-deps
 
 ARG ROCKY_NGEN_DEPS_REQUIRED
+ARG NUMPY_CONSTRAINT
 
 # TODO: later, go back and change all pip3/python3 to just pip/python (but leave for now to limit scope)
 # Note that this includes numpy, which is needed for Python BMI support, regardless of BMI module 
 USER root
```

I figure we could move the _required_ python packages to a `requirements.txt` file (so, `pip`, `wheel`, `packaging`, and `numpy`) and then create a `constraints.txt` that we use to enforce the allowed versions of the required and other packages in `constraints.txt`. Thoughts?

```suggestion
USER root
COPY requirements.txt ${WORKDIR}/requirements.txt
COPY constraints.txt ${WORKDIR}/constraints.txt
```

---

Good find! I think it's a good solution.

It's been a while since I've had to worry about this, but I _believe_ that editing the `constraints.txt` and creating a new hash for the file would cause a new `docker build` to start at the line for the `constraints.txt` file , so I think we can be sure that a change there will just about always be reflected after a build.

---

I still don't like the idea of coupling all the Python packages together in a _requirements.txt_, as I don't think that fits with how the Dockerfile is set up.  It's totally fair to discuss whether _that_ needs to change - e.g., should we _really_ support building the image both with or without Python support (we don't really need numpy if in the without option)) - but it needs to be discussed separately.

But the `PIP_CONSTRAINT` and _constraints.txt_ settings sound fine.  Those alone should be sufficient to resolve the immediate issue.

---

@christophertubbs, you raise a good point I hadn't consider:  depending on when the _constraints.txt_ file is setup in the image that will once again have implications on build caching and efficiency.  That isn't an absolute blocker, but I want to think on it a little more to see if there's anyway to avoid that tradeoff.

---

> I still don't like the idea of coupling all the Python packages together in a _requirements.txt_, as I don't think that fits with how the Dockerfile is set up.

Yeah, I agree. Im just talking about the required packages to build `ngen` with python support. So, `numpy` for now but this could expand to others in the future.